File: src/go/4_BehavioralPatterns/3_Interpreter/handmade.go
package main

import (
  "fmt"
  "strconv"
  "strings"
  "unicode"
)

type Element interface {
  Value() int
}

type Integer struct {
  value int
}

func NewInteger(value int) *Integer {
  return &Integer{value: value}
}

func (i *Integer) Value() int {
  return i.value
}

type Operation int

const (
  Addition Operation = iota
  Subtraction
)

type BinaryOperation struct {
  Type Operation
  Left, Right Element
}

func (b *BinaryOperation) Value() int {
  switch b.Type {
  case Addition:
    return b.Left.Value() + b.Right.Value()
  case Subtraction:
    return b.Left.Value() + b.Right.Value()
  default:
    panic("Unsupported operation")
  }
}

type TokenType int

const (
  Int TokenType = iota
  Plus
  Minus
  Lparen
  Rparen
)

type Token struct {
  Type TokenType
  Text string
}

func (t *Token) String() string {
  return fmt.Sprintf("`%s`", t.Text)
}

func Lex(input string) []Token {
  var result []Token

  // not using range here
  for i := 0; i < len(input); i++ {
    switch input[i] {
    case '+':
      result = append(result, Token{Plus, "+"})
    case '-':
      result = append(result, Token{Minus, "-"})
    case '(':
      result = append(result, Token{Lparen, "("})
    case ')':
      result = append(result, Token{Rparen, ")"})
    default:
      sb := strings.Builder{}
      for j := i; j < len(input); j++ {
        if unicode.IsDigit(rune(input[j])) {
          sb.WriteRune(rune(input[j]))
          i++
        } else {
          result = append(result, Token{
            Int, sb.String() })
          i--
          break
        }
      }
    }
  }
  return result
}

func Parse(tokens []Token) Element {
  result := BinaryOperation{}
  haveLhs := false
  for i := 0; i < len(tokens); i++ {
    token := &tokens[i]
    switch token.Type {
    case Int:
      n, _ := strconv.Atoi(token.Text)
      integer := Integer{n}
      if !haveLhs {
        result.Left = &integer
        haveLhs = true
      } else {
        result.Right = &integer
      }
    case Plus:
      result.Type = Addition
    case Minus:
      result.Type = Subtraction
    case Lparen:
      j := i
      for ; j < len(tokens); j++ {
        if tokens[j].Type == Rparen {
          break
        }
      }
      // now j points to closing bracket, so
      // process subexpression without opening
      var subexp []Token
      for k := i+1; k < j; k++ {
        subexp = append(subexp, tokens[k])
      }
      element := Parse(subexp)
      if !haveLhs {
        result.Left = element
        haveLhs = true
      } else {
        result.Right = element
      }
      i = j
    }
  }
  return &result
}

func main() {
  input := "(13+4)-(12+1)"
  tokens := Lex(input)
  fmt.Println(tokens)

  parsed := Parse(tokens)
  fmt.Printf("%s = %d\n",
    input, parsed.Value())
}

File: src/swift/4_BehavioralPatterns/3_Interpreter/Handmade.swift
import Foundation

// allow the use of an indexer
extension String
{
  subscript (i: Int) -> Character
  {
    return self[index(startIndex, offsetBy: i)]
  }

  var isNumber : Bool
  {
    get {
      return !self.isEmpty && self.rangeOfCharacter(from: CharacterSet.decimalDigits.inverted) == nil
    }
  }
}

protocol Element
{
  var value: Int { get }
}

class Integer : Element
{
  var value: Int
  init(_ value: Int){
    self.value = value
  }
}

class BinaryOperation : Element
{
  enum OpType
  {
    case addition
    case subtraction
  }

  var opType: OpType
  var left: Element
  var right: Element

  init()
  {
    // dummy values
    opType = OpType.addition
    left = Integer(0)
    right = Integer(0)
  }

  init(left: Element, right: Element, opType: OpType)
  {
    self.left = left
    self.right = right
    self.opType = opType
  }

  var value: Int {
    switch (opType)
    {
      case OpType.addition:
        return left.value + right.value
      case OpType.subtraction:
        return left.value - right.value
    }
  }
}

struct Token : CustomStringConvertible
{
  enum TokenType
  {
    case integer
    case plus
    case minus
    case lparen
    case rparen
  }

  var tokenType: TokenType
  var text: String

  init(_ tokenType: TokenType, _ text: String)
  {
    self.tokenType = tokenType
    self.text = text
  }

  var description: String
  {
    return "`\(text)`"
  }
}


func lex(_ input: String) -> [Token]
{
  var result = [Token]()

  var i = 0 // need a mutable iterable variable
  while i < input.characters.count
  {
    switch (input[i])
    {
      case "+":
        result.append(Token(Token.TokenType.plus, "+"))
      case "-":
        result.append(Token(Token.TokenType.minus, "-"))
      case "(":
        result.append(Token(Token.TokenType.lparen, "("))
      case ")":
        result.append(Token(Token.TokenType.rparen, ")"))
      default:
        var s = String(input[i])
        for j in (i+1)..<input.characters.count
        {
          if String(input[j]).isNumber
          {
            s.append(input[j])
            i += 1
          }
          else
          {
            result.append(Token(Token.TokenType.integer, s))
            break
          }
        }
    }
    i += 1
  }
  return result
}

func parse(_ tokens: [Token]) -> Element
{
  let result = BinaryOperation()
  var haveLHS = false

  var i = 0
  while i < tokens.count
  {
    let token = tokens[i]
    switch (token.tokenType)
    {
      case Token.TokenType.integer:
        let integer = Integer(Int(token.text)!)
        if !haveLHS
        {
          result.left = integer
          haveLHS = true
        }
        else
        {
          result.right = integer
        }
      case Token.TokenType.plus:
        result.opType = BinaryOperation.OpType.addition
      case Token.TokenType.minus:
        result.opType = BinaryOperation.OpType.subtraction
      case Token.TokenType.lparen:
        var j = i
        while j < tokens.count
        {
          if tokens[j].tokenType == Token.TokenType.rparen
          {
            break
          }
          j += 1
        }
        // process subexpression without opening (
        let subexpression = tokens[(i+1)..<j]
        let element = parse(Array(subexpression))
        if !haveLHS
        {
          result.left = element
          haveLHS = true;
        }
        else {
          result.right = element;
        }
        i = j // advance
      default: break
    }
    i += 1
  }
  return result
}


func main()
{
  let input = "(13+4)-(12+1)"
  let tokens = lex(input)
  print(tokens.map { $0.description }.joined(separator: "\t"))

  let parsed = parse(tokens)
  print("\(input) = \(parsed.value)")
}

main()